mantidproject/mantid
# Allow for nested dict inputs for Python algorithms

Recently a `dict` input was provided for Python Algorithms. These dictionary inputs are transformed into `PropertyManager` objects. This is a very convenient and clean way of providing a large set of input parameters to an algorithm.

Up until now nested `dict`s could not be handled. This PR attempts to allow for nested `dict`s as inputs to Python algorithms.

**To test:**

Pleas find the Ubuntu installer here: \olympic\Babylon5\Scratch\Anton\Testing\16665_Python_Dicts
- Run the following Python script and confirm that the nested dict inputs have translated correctly into nested property managers. 

``` python
class FakeAlgorithm(Algorithm):
    def PyInit(self):
        self.declareProperty(PropertyManagerProperty("Args"))

    def PyExec(self):
        pass

def evaluate_pmgr(pmgr, key, expected):
    print "For {} expected a value of {} and received a value of {}".format(key, pmgr[key].value, expected)

# ###################
# Test case
# ###################
fake = FakeAlgorithm()
fake.initialize()
fake.setProperty("Args", {'A': 1, 
                                     'B':10.5, 
                                     'C':'String arg', 
                                     'D': [0.0,11.3], 
                                     'E':{'F':10.4, 
                                            'G': [1.0,2.0, 3.0],
                                             'H':{'I': "test",
                                                     'J': 120.6}}})

pmgr = fake.getProperty("Args").value
evaluate_pmgr(pmgr, "A", 1)
evaluate_pmgr(pmgr, "B", 10.5)
evaluate_pmgr(pmgr, "C", 'String arg')
evaluate_pmgr(pmgr, "D", [0.0,11.3])

# Get the level1-nested property manager 
nested_l1_pmgr = pmgr["E"].value
evaluate_pmgr(nested_l1_pmgr, "F", 10.4)
evaluate_pmgr(nested_l1_pmgr, "G", [1.0,2.0, 3.0])

# Get the level2-nested property manager 
nested_l2_pmgr = nested_l1_pmgr["H"].value
evaluate_pmgr(nested_l2_pmgr, "I", "test")
evaluate_pmgr(nested_l2_pmgr, "J", 120.6)
```
- If you can think of other `dict` scenarios try them on `FakeAlgorithm`

Fixes #16667

_Does not need to be in the release notes._

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
##### Functional Tests
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
